packages/mobx-state-tree/__tests__/core/model.md

```diff
@@ -0,0 +1,129 @@
+
+1. A set of properties that describe the tree and its possible state.
+2. A function called `views`, which will allow us to set up and apply [views](https://mobx-state-tree.js.org//intro/getting-started#model-views) to the instance.
+3. A function called `actions`, which will allow us to set up [actions](https://mobx-state-tree.js.org/concepts/actions) that can modify the instance and its subtree(s)
+4. A function called `volatile`, which will allow us to set up [volatile](https://mobx-state-tree.js.org/concepts/volatiles) state for the object.
+5. An `extend` method, which allows us to [share state between views and actions](https://github.com/mobxjs/mobx-state-tree/pull/425/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R630).
+6. Methods called `preProcessSnapshot` and `postProcessSnapshot`, which can [transform snapshots or take action after processing snapshots for the instance](https://mobx-state-tree.js.org/API/interfaces/isnapshotprocessors#methods-1).
+
+## How we get there
+
+To keep this post tightly scoped, I'm going to focus on what goes on in the [`ModelType` constructor](https://github.com/mobxjs/mobx-state-tree/blob/048f1ebf0d2c49983ec8914d78162882bc1a3e3e/packages/mobx-state-tree/src/types/complex-types/model.ts#L342):
+
+1. We set the name of the model.
+2. We parse the provided properties into an actual instance of [`ModelProperties`](https://github.com/mobxjs/mobx-state-tree/blob/048f1ebf0d2c49983ec8914d78162882bc1a3e3e/packages/mobx-state-tree/src/types/complex-types/model.ts#L65) (notice how we took an `ModelPropertiesDeclaration` and turned it into an actual `ModelProperties`).
+3. We freeze the properties (another internal MST function that deserves its own post).
+4. We determine the identifier attribute if there is one.
+
+Here's what that looks like:
+
+```ts
+ constructor(opts: ModelTypeConfig) {
+        super(opts.name || defaultObjectOptions.name)
+        Object.assign(this, defaultObjectOptions, opts)
+        // ensures that any default value gets converted to its related type
+        this.properties = toPropertiesObject(this.properties) as PROPS
+        freeze(this.properties) // make sure nobody messes with it
+        this.propertyNames = Object.keys(this.properties)
+        this.identifierAttribute = this._getIdentifierAttribute()
+    }
+```
+
+## Writing some tests
+
+I hope you're feeling pretty good about the high level overview so far. This helped me to map out what's going on with my models in MobX-State-Tree. Let's see it in action! I want to write some tests around the behavior I've described.
+
+### Testing names
+
+1. Providing a string as the first argument should set it as the model's name.
+2. Providing an empty string as the first argument should set it as the model's name.
+3. Providing a non-string argument as the first argument should set the model's name as 'AnonymousModel'.
+
+### Testing properties
```

Comment:
When I pull this out to my own blog, I'll add a link to this PR or the test file I updated.